File: python/exercises/practice/unit_test_a49e025e_fdd7_41cd_b02c_4f0ac37aac64/transaction_manager/.meta/transaction_manager.py
import threading
from enum import Enum

class TransactionStatus(Enum):
    PENDING = "PENDING"
    COMMITTED = "COMMITTED"
    ROLLED_BACK = "ROLLED_BACK"
    NON_EXISTENT = "NON_EXISTENT"

class TransactionManager:
    def __init__(self):
        self.transactions = {}
        self.lock = threading.RLock()

    def begin_transaction(self, transaction_id):
        with self.lock:
            if transaction_id in self.transactions:
                return False
            
            self.transactions[transaction_id] = {
                'status': TransactionStatus.PENDING,
                'participants': set(),
                'committed': set(),
                'rolled_back': set()
            }
            return True

    def register_participant(self, transaction_id, service_id):
        with self.lock:
            tx = self.transactions.get(transaction_id)
            if not tx or tx['status'] != TransactionStatus.PENDING:
                return False
            
            if service_id in tx['participants']:
                return False
                
            tx['participants'].add(service_id)
            return True

    def service_commit(self, transaction_id, service_id):
        with self.lock:
            tx = self.transactions.get(transaction_id)
            if not tx or tx['status'] != TransactionStatus.PENDING:
                return False
                
            if service_id not in tx['participants']:
                return False
                
            if service_id in tx['committed'] or service_id in tx['rolled_back']:
                return False
                
            tx['committed'].add(service_id)
            
            if tx['committed'] == tx['participants']:
                tx['status'] = TransactionStatus.COMMITTED
                
            return True

    def service_rollback(self, transaction_id, service_id):
        with self.lock:
            tx = self.transactions.get(transaction_id)
            if not tx or tx['status'] != TransactionStatus.PENDING:
                return False
                
            if service_id not in tx['participants']:
                return False
                
            if service_id in tx['committed'] or service_id in tx['rolled_back']:
                return False
                
            tx['rolled_back'].add(service_id)
            tx['status'] = TransactionStatus.ROLLED_BACK
            return True

    def get_transaction_status(self, transaction_id):
        with self.lock:
            tx = self.transactions.get(transaction_id)
            if not tx:
                return TransactionStatus.NON_EXISTENT.value
            return tx['status'].value

File: java/exercises/practice/unit_test_6606ec59_d308_4d32_9eea_43f399f49e82/distributed_tx/.meta/src/main/java/TransactionManager.java
package distributed_tx;

import java.util.*;
import java.util.concurrent.*;

public class TransactionManager {

    private final Map<String, Service> services;
    private final Map<String, Transaction> transactions;
    private final ExecutorService executor;
    private final long prepareTimeoutMillis;

    public TransactionManager() {
        this.services = new ConcurrentHashMap<>();
        this.transactions = new ConcurrentHashMap<>();
        this.executor = Executors.newCachedThreadPool();
        this.prepareTimeoutMillis = 2000; // default timeout of 2000ms for prepare phase
    }

    public String beginTransaction() {
        String txId = UUID.randomUUID().toString();
        Transaction tx = new Transaction(txId);
        transactions.put(txId, tx);
        return txId;
    }

    public void registerService(Service service) {
        services.put(service.getName(), service);
    }

    public void addOperation(String transactionId, Operation op) {
        Transaction tx = transactions.get(transactionId);
        if (tx == null) {
            throw new IllegalArgumentException("Transaction not found: " + transactionId);
        }
        if (!services.containsKey(op.getServiceName())) {
            throw new IllegalArgumentException("Service not registered: " + op.getServiceName());
        }
        tx.addOperation(op);
    }

    public void executeTransaction(String transactionId) {
        Transaction tx = transactions.get(transactionId);
        if (tx == null) {
            throw new IllegalArgumentException("Transaction not found: " + transactionId);
        }
        synchronized (tx) {
            if (tx.getStatus() != TransactionStatus.PENDING) {
                // Transaction already executed; idempotent execution.
                return;
            }
            List<Operation> ops = tx.getOperations();
            Set<String> participantNames = new HashSet<>();
            for (Operation op : ops) {
                participantNames.add(op.getServiceName());
            }
            List<Service> participants = new ArrayList<>();
            for (String name : participantNames) {
                Service service = services.get(name);
                if (service != null) {
                    participants.add(service);
                }
            }

            boolean prepareSuccess = true;
            // Phase 1: Prepare
            for (Operation op : ops) {
                Service service = services.get(op.getServiceName());
                try {
                    Future<Boolean> future = executor.submit(() -> service.prepare(transactionId, op));
                    boolean result = future.get(prepareTimeoutMillis, TimeUnit.MILLISECONDS);
                    if (!result) {
                        prepareSuccess = false;
                        break;
                    }
                } catch (Exception e) {
                    prepareSuccess = false;
                    break;
                }
            }

            // Phase 2: Commit or Rollback based on prepare outcome.
            if (prepareSuccess) {
                for (Service service : participants) {
                    try {
                        service.commit(transactionId);
                    } catch (Exception e) {
                        // In a full system, additional recovery logic would be applied.
                    }
                }
                tx.setStatus(TransactionStatus.COMMITTED);
            } else {
                for (Service service : participants) {
                    try {
                        service.rollback(transactionId);
                    } catch (Exception e) {
                        // In a full system, additional recovery/retry logic would be applied.
                    }
                }
                tx.setStatus(TransactionStatus.ROLLEDBACK);
            }
        }
    }

    public TransactionStatus getTransactionStatus(String transactionId) {
        Transaction tx = transactions.get(transactionId);
        if (tx == null) {
            throw new IllegalArgumentException("Transaction not found: " + transactionId);
        }
        return tx.getStatus();
    }

    // Internal transaction representation.
    private static class Transaction {
        private final String transactionId;
        private final List<Operation> operations;
        private TransactionStatus status;

        public Transaction(String transactionId) {
            this.transactionId = transactionId;
            this.operations = new ArrayList<>();
            this.status = TransactionStatus.PENDING;
        }

        public void addOperation(Operation op) {
            operations.add(op);
        }

        public List<Operation> getOperations() {
            return operations;
        }

        public TransactionStatus getStatus() {
            return status;
        }

        public void setStatus(TransactionStatus status) {
            this.status = status;
        }
    }
}